src/v/cluster/controller_backend.cc

```diff
@@ -741,58 +752,74 @@ ss::future<> controller_backend::clear_orphan_topic_files(
       });
 }
 
-void controller_backend::start_topics_reconciliation_loop() {
+void controller_backend::start_fetch_deltas_loop() {
     ssx::spawn_with_gate(_gate, [this] {
         return ss::do_until(
           [this] { return _as.local().abort_requested(); },
           [this] {
               return fetch_deltas()
-                .then([this] { return reconcile_topics(); })
                 .handle_exception_type(
                   [](const ss::abort_requested_exception&) {
                       // Shutting down: don't log this exception as an error
                       vlog(
                         clusterlog.debug,
-                        "Abort requested while reconciling topics");
+                        "Abort requested while fetching deltas");
                   })
                 .handle_exception([](const std::exception_ptr& e) {
                     vlog(
-                      clusterlog.error,
-                      "Error while reconciling topics - {}",
-                      e);
+                      clusterlog.error, "Error while fetching deltas - {}", e);
                 });
           });
     });
 }
 
-void controller_backend::housekeeping() {
-    ssx::background
-      = ssx::spawn_with_gate_then(_gate, [this] {
-            auto f = ss::now();
-            if (!_states.empty() && _topics_sem.available_units() > 0) {
-                f = reconcile_topics();
+ss::future<> controller_backend::reconcile_ntp_fiber(
+  model::ntp ntp, ss::lw_shared_ptr<ntp_reconciliation_state> rs) {
+    if (_gate.is_closed()) {
+        co_return;
+    }
+    auto gate_holder = _gate.hold();
+
+    while (!_as.local().abort_requested()) {
+        try {
+            co_await rs->wakeup_signal.wait(_housekeeping_timer_interval);
```

Comment:
this part seems a little bit fishy for me. i.e. we are waiting with timeout and then ignoring it. Can we sleep if error was returned from `try_reconcile_ntp` and there is still work to do and here check the predicate indicating the the ntp needs reconciliation ?

Reply:
The timed out condition wait is precisely sleeping for the full housekeeping interval, but we want to wake up from it if new delta has arrived (e.g. we were waiting to retry replicas update, but an update_finished command arrived, then we can make progress immediately). So right now the logic is:
1. do reconciliation work until finished or error encountered
2. if finished, delete the state from the map and exit
3. else wait on the condition for the sleep period to elapse or for the new wakeup signal to arrive
4. repeat